semgrep-core/src/matching/Generic_vs_generic.ml

```diff
@@ -1357,15 +1357,19 @@ and type_of_expr e : (G.ident * G.type_ option) option =
   (* deep: same *)
   | B.Call
       ( { e = B.DotAccess (_, _, FN (Id (idb, { B.id_type = tb; _ }))); _ },
-        _args ) -> (
+        _args )
+  (* deep: in Java, there can be an implicit `this.`
+     so look this up the type here too *)
+  | B.Call ({ e = N (Id (idb, { B.id_type = tb; _ })); _ }, _args)
+    when lang = Lang.Java -> (
```

Comment:
This `when` clause will apply to every part of the pattern match that uses this case. It doesn't sound like that was your intent? If so, could you add a clarification comment?

```
let f z = function
  | 1
  | 2 when z -> print_endline "first"
  | _ -> print_endline "second"
;;

f false 1
```
prints "second"